lib/src/gitattributes.rs

```diff
@@ -0,0 +1,705 @@
+            ),
+            (
+                "foo2/.gitattributes",
+                "# this macro definition shouldn't take effect.
+                    [attr]override_macro d
+                    # this macro definition shouldn't take effect.
+                    [attr]new_macro e
+                    bar base_macro override_macro new_macro",
+            ),
+        ]);
+        assert_search_output!(&git_attributes, "foo1/bar", "a", State::Set);
+        assert_search_output!(&git_attributes, "foo1/bar", "b", State::Set);
+        assert_search_output!(&git_attributes, "foo1/bar", "c", State::Unspecified);
+        assert_search_output!(&git_attributes, "foo1/bar", "d", State::Unspecified);
+        assert_search_output!(&git_attributes, "foo1/bar", "e", State::Unspecified);
+        assert_search_output!(&git_attributes, "foo1/bar", "new_macro", State::Set);
+
+        assert_search_output!(&git_attributes, "foo2/bar", "a", State::Set);
+        assert_search_output!(&git_attributes, "foo2/bar", "b", State::Set);
+        assert_search_output!(&git_attributes, "foo2/bar", "c", State::Unspecified);
+        assert_search_output!(&git_attributes, "foo2/bar", "d", State::Unspecified);
+        assert_search_output!(&git_attributes, "foo2/bar", "e", State::Unspecified);
+        assert_search_output!(&git_attributes, "foo2/bar", "new_macro", State::Set);
+    }
+
+    #[test]
+    fn search_priority_and_fallback() {
+        let disk: TestStore = [("foo/.gitattributes", "*.txt text")]
+            .iter()
+            .map(|(key, value)| {
+                (
+                    RepoPathBuf::from_internal_string(key.to_string()).unwrap(),
+                    Ok(value.to_string()),
+                )
+            })
+            .collect();
+
+        let store: TestStore = [("bar/.gitattributes", "*.txt text")]
+            .iter()
+            .map(|(key, value)| {
+                (
+                    RepoPathBuf::from_internal_string(key.to_string()).unwrap(),
+                    Ok(value.to_string()),
+                )
+            })
+            .collect();
+
+        let git_attributes = GitAttributes::new(store, disk);
+        assert_search_output!(
+            &git_attributes,
+            "foo/bar.txt",
+            "text",
+            State::Set,
+            SearchPriority::Disk
+        );
+        assert_search_output!(
+            &git_attributes,
+            "foo/bar.txt",
+            "text",
+            State::Set,
+            SearchPriority::Store
+        );
+        assert_search_output!(
+            &git_attributes,
+            "bar/bar.txt",
+            "text",
+            State::Set,
+            SearchPriority::Store
+        );
+        assert_search_output!(
+            &git_attributes,
+            "bar/bar.txt",
+            "text",
+            State::Set,
+            SearchPriority::Disk
+        );
+    }
+
+    #[test]
+    fn file_loader_io_error() {
```

Comment:
According to my sampling from `jj`, almost all the tests are prefixed with `test_`. Can we rename the test function with `test_` prefixed?